### DISCUSSION TITLE
How should I go about learning winnow?

### DISCUSSION THREAD

**Discussion by bigbass1997** [winnow-rs/winnow#632] [Q&A]
While struggling to learn how to use `winnow` I'm finding that I have many questions that aren't answered by reading the docs. Every time a question pops up, I inevitably spend quite a lot of time wandering between the docs, github discussions and issues, and google, looking for help. Though I rarely find any.

<hr>

For example, in attempting to write a parser, I needed to match against several different possible characters at the start of a string to determine how to parse the rest.

I figured a `match` would probably work. Thinking that maybe this would be common enough for `winnow` to provide a parser for this scenario, I searched the [parser list](https://docs.rs/winnow/latest/winnow/combinator/index.html) and found that `dispatch!` appeared to be what I wanted. However, several different questions surfaced as a result.

Looking at the `dispatch!` example, I saw `empty.value()`. What is [empty](https://docs.rs/winnow/latest/winnow/combinator/fn.empty.html)? Perhaps this is a tangent, but in order to understand how to use dispatch, I expected I needed to understand the example in full.

According to the list of parsers:
> `empty`: Returns a value without consuming any input, always succeeds

How can "empty" return a value? Why is it called empty if it's not actually empty? The docs for the function didn't explain it at all. Simply saying "Succeed, consuming no input", referring me to other parsers and an example which didn't offer any obvious explanation. (In fact, the example ended up introducing even more stuff I didn't know about, like `IResult` and `parse_peek`)

I guessed that `value()` must be where the value comes from. But why bother using `empty` in the first place? Also, if `empty.value('c')` returns the value `'c'` why not just use `'c'` by itself instead of the full `empty.value('c')`?

<hr>

After literally hours of falling down rabbit holes, I sorta understand some of this on a surface level, but my questions are still unanswered. The whole endeavor leaves me feeling demotivated.

I think there's two fundamental issues I'm having:
1. In order to learn how and why I should use one piece of the crate (e.g. a parser), I feel forced to decipher examples which introduce several other unknown concepts/parsers. Resulting in a never-ending recursive loop of foreign concepts.
2. It is incredibly difficult to reason about _how_ to create a complete parser. What steps should I take? Where should I begin?
   * I can look at a [full example](https://docs.rs/winnow/latest/winnow/_topic/ini/index.html) and roughly understand what the code is probably doing. But, how did the author of that code know which parsers/combinators to use?

Note, I have read the Tutorial in the docs, multiple times. It does a fine job of giving simple examples of text to parse, and telling you precisely what to use to parse it. But as soon as I have to parse something it didn't explicitly mention, I find myself completely lost.

**Are there any other resources, besides the rustdocs, for learning how to use `winnow` to write parsers?** Perhaps something that walks through the _thought process_ behind writing a parser?

Apologies if I've missed anything obvious, or haven't explained something well enough.

**Comment by bigbass1997** [comment]
After resting on this overnight, I think I might be starting to understand more of the crate. Is using winnow essentially like rolling your own scanning parser, but factoring out common checks into separate parsers? (parsers that can also be readily combined)

So instead of checking every token by hand, with a web of "if this, then _take_ tokens _until_ this, or else do this _alternative_", winnow provides a bunch of combinators/parsers that take the place of all the common actions?

I still have a lot of questions that mostly boil down to "why use <insert parser>?" or other small questions, like "is there a parser that does X?"

**Reply by epage** [reply]
> Is using winnow essentially like rolling your own scanning parser, but factoring out common checks into separate parsers? (parsers that can also be readily combined)

Yes.  This isn't about defining BNF and having things automatically figured out but how to process an input, left to right, like you would when hand writing a parser.  

In ["why winnow?"](https://docs.rs/winnow/latest/winnow/_topic/why/index.html#winnow), we describe it as:

> general version of the helpers you would create along the way to writing a hand-written parser.

[Chapter 1](https://docs.rs/winnow/latest/winnow/_tutorial/chapter_1/index.html) of the tutorial also attempts to pain this picture and if there are ideas on how to better communicate that, we'd love to hear your thoughts!

> I still have a lot of questions that mostly boil down to "why use ?" or other small questions, like "is there a parser that does X?"

The tutorial also attempts to give you the principles for how to compose things so you are better prepared to make whatever parser you want.

For example, [Repetition](https://docs.rs/winnow/latest/winnow/_tutorial/chapter_5/index.html) also introduces `Parser::take` and how you can use that to capture of slice of arbitrary content.

**Comment by epage** [comment]
Sorry for the delay as I'm still working to catch up for US holidays.  There is a lot here, so I'll try to break up each thread by topic.

## `dispatch!`

> Looking at the dispatch! example, I saw empty.value(). What is [empty](https://docs.rs/winnow/latest/winnow/combinator/fn.empty.html)? Perhaps this is a tangent, but in order to understand how to use dispatch, I expected I needed to understand the example in full.

There tends to be tension with examples with being standalone vs being practical.

> I guessed that value() must be where the value comes from. But why bother using empty in the first place?

`dispatch` maps the output of a parser to a parser, not a value.  So if the branch's parser was able to make a decision purely off of the matched parser, then `empty` works.

**Comment by epage** [comment]
## `empty`


> > According to the list of parsers:
>
> empty: Returns a value without consuming any input, always succeeds

That looks like its stale from when `empty` was `value(n)` which was the equivalent of `empty.value(n)`.

I would also recommend to keep in mind that the list of parsers is an index and not complete on its own.  I'd recommend following the link for more details.

> Why is it called empty if it's not actually empty? 

We try to name parsers after their grammatical significance.  This parser is a grammar no-op / an empty slice of the input.  I actually wonder if we should change the return value to return a slice by default.

If you have an idea for a better name, we are open to it!  Names make a big difference in understandability and we're always looking to improve on that front.

> referring me to other parsers 

We're trying to help you find relevant building blocks for what you are trying to accomplish.

>  (In fact, the example ended up introducing even more stuff I didn't know about, like IResult and parse_peek)

With 0.7, `IResult` has been removed.  `parse_peek` is still used throughout our examples though.

**Comment by epage** [comment]
FYI I've posted #756.
